### DISCUSSION TITLE
QST:sjoin_nearest can't be used

### DISCUSSION THREAD

**Discussion by lx12633036** [geopandas/geopandas#2717] [Q&A]
- [x] I have searched the [geopandas] tag on [StackOverflow](https://stackoverflow.com/questions/tagged/geopandas) and [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/geopandas) for similar questions.

- [ ] I have asked my usage related question on [StackOverflow](https://stackoverflow.com) or [GIS StackExhange](https://gis.stackexchange.com).

---

#### Question about geopandas

**Note**: If you'd still like to submit a question, please read [this guide](
https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your question.

```python
# Your code here, if applicable
[NotImplementedError: Currently, only PyGEOS >= 0.10.0 supports `nearest_all`. To use PyGEOS within GeoPandas, you need to install PyGEOS: 'conda install pygeos' or 'pip install pygeos']
```
But，my geopandas version is 0.10.2. And my pygeos version is 0.12.0

**Comment by martinfleis** [comment]
Hi, are you sure you are running the code in the same environment where you have pygeos? Can you post what `geopandas.show_versions()` returns in that environment?

**Comment by lx12633036** [comment]
> Hi, are you sure you are running the code in the same environment where you have pygeos? Can you post what `geopandas.show_versions()` returns in that environment?

Hello，thank you for your reply!
This is my  environment.In this environment, I can‘t see pygeos，is that right?
SYSTEM INFO
-----------
python     : 3.8.12 (default, Oct 12 2021, 03:01:40) [MSC v.1916 64 bit (AMD64)]

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.3.3
GDAL data dir: None
PROJ       : 8.2.0

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.2
pandas     : 1.4.1
fiona      : 1.8.20
numpy      : 1.22.2
shapely    : 1.8.0
rtree      : 0.9.7
pyproj     : 3.3.0
matplotlib : 3.5.0
mapclassify: 2.4.3
geopy      : None
psycopg2   : None
geoalchemy2: None
pyarrow    : None
pygeos     : None

**Comment by martinfleis** [comment]
It seems that you do not have `pygeos` installed in your environment. Can you double check that?

**Comment by lx12633036** [comment]
should i  need to import pygeos separately

**Comment by jorisvandenbossche** [comment]
There is no need to import pygeos separately, but you do need to install in the environment. From the output of `geopandas.show_versions()`, you can see that pygeos is not installed in that environment (`None` means not installed).

**Comment by lx12633036** [comment]
But in my pip list, I can see pygeos 0.12.0 in my list! 
How can i import pygeos?

**Comment by lx12633036** [comment]
`gpd.options.use_pygeos=True`
I use this in my notebook,but an ERROR is
`[ImportError: To use PyGEOS within GeoPandas, you need to install PyGEOS: 'conda install pygeos' or 'pip install pygeos']()`
this is my pip list:
`Package                   Version
------------------------- -----------
affine                    2.3.0
alembic                   1.4.1
appdirs                   1.4.4
argon2-cffi               21.3.0
argon2-cffi-bindings      21.2.0
arviz                     0.11.4
asttokens                 2.0.5
atomicwrites              1.4.0
attrs                     21.2.0
aws-requests-auth         0.4.3
backcall                  0.2.0
backports.zoneinfo        0.2.1
beautifulsoup4            4.10.0
black                     21.12b0
bleach                    4.1.0
blis                      0.7.5
Boruta                    0.3
boto3                     1.15.18
botocore                  1.18.18
bs4                       0.0.1
cachetools                4.2.4
catalogue                 2.0.6
certifi                   2021.10.8
cffi                      1.15.0
cftime                    1.5.1.1
charset-normalizer        2.0.8
click                     8.0.3
click-plugins             1.1.1
cligj                     0.7.2
cloudpickle               2.0.0
colorama                  0.4.4
colorlover                0.3.0
contextily                1.2.0
CoordinatesConverter      0.1.4
coverage                  6.2
cufflinks                 0.17.3
cycler                    0.11.0
cymem                     2.0.6
Cython                    0.29.14
databricks-cli            0.16.2
debugpy                   1.5.1
decorator                 5.1.0
defusedxml                0.7.1
Deprecated                1.2.13
descartes                 1.1.0
dill                      0.3.4
dnspython                 2.1.0
docker                    5.0.3
docopt                    0.6.2
entrypoints               0.3
esda                      2.4.1
et-xmlfile                1.1.0
executing                 0.8.2
fastprogress              1.0.0
filelock                  3.4.0
Fiona                     1.8.20
Flask                     2.0.2
Flask-Cors                3.0.10
Flask-JSON                0.3.4
fonttools                 4.28.2
funcy                     1.17
future                    0.18.2
GDAL                      3.3.3
gensim                    3.8.3
geographiclib             1.52
geopandas                 0.10.2
geopy                     2.2.0
gitdb                     4.0.9
GitPython                 3.1.24
greenlet                  1.1.2
html5lib                  1.1
htmlmin                   0.1.12
idna                      3.3
ImageHash                 4.2.1
imbalanced-learn          0.9.0
importlib-metadata        4.11.2
iniconfig                 1.1.1
ipykernel                 6.9.1
ipython                   8.0.1
ipython-genutils          0.2.0
ipywidgets                7.6.5
itsdangerous              2.0.1
jedi                      0.18.1
Jinja2                    3.0.3
jmespath                  0.10.0
joblib                    1.1.0
Js2Py                     0.71
jsonlines                 3.0.0
jsonschema                3.2.0
jupyter-client            7.1.0
jupyter-core              4.9.1
jupyterlab-pygments       0.1.2
jupyterlab-widgets        1.0.2
kiwisolver                1.3.2
kmodes                    0.11.1
langcodes                 3.3.0
Levenshtein               0.18.1
libpysal                  4.6.2
lightgbm                  3.3.1
llvmlite                  0.37.0
lxml                      4.8.0
Mako                      1.1.6
mapclassify               2.4.3
MarkupSafe                2.1.0
matplotlib                3.5.0
matplotlib-inline         0.1.3
mercantile                1.2.1
missingno                 0.5.0
mistune                   0.8.4
mlflow                    1.22.0
mlxtend                   0.19.0
mpmath                    1.2.1
multimethod               1.6
munch                     2.5.0
murmurhash                1.0.6
mypy-extensions           0.4.3
nbclient                  0.5.9
nbconvert                 6.3.0
nbformat                  5.1.3
nest-asyncio              1.5.1
netCDF4                   1.5.8
networkx                  2.7.1
nltk                      3.6.7
notebook                  6.4.6
numba                     0.54.1
numexpr                   2.8.1
numpy                     1.22.2
openpyxl                  3.0.9
packaging                 21.3
palettable                3.3.0
pandas                    1.4.1
pandas-profiling          3.1.0
pandocfilters             1.5.0
parso                     0.8.2
pathspec                  0.9.0
pathy                     0.6.1
patsy                     0.5.2
phik                      0.12.0
pickleshare               0.7.5
Pillow                    9.0.1
pinyin                    0.4.0
pip                       21.3.1
pipwin                    0.5.1
plac                      1.1.3
platformdirs              2.4.1
plot-map                  0.3.6
plotly                    5.6.0
pluggy                    1.0.0
preshed                   3.0.6
prometheus-client         0.12.0
prometheus-flask-exporter 0.18.7
prompt-toolkit            3.0.23
protobuf                  3.19.1
pure-eval                 0.2.1
py                        1.11.0
pycaret                   2.3.5
pycparser                 2.21
pydantic                  1.8.2
pygeos                    0.12.0
Pygments                  2.10.0
pyjsparser                2.7.1
pyLDAvis                  3.2.2Note: you may need to restart the kernel to use updated packages.

pymc3                     3.11.4
pynndescent               0.5.5
pyod                      0.9.6
pyparsing                 3.0.6
PyPrind                   2.11.3
pyproj                    3.3.0
pyrsistent                0.18.0
pysal                     2.1.0
pySmartDL                 1.3.4
python-dateutil           2.8.2
python-editor             1.0.4
pytz                      2021.3
pytz-deprecation-shim     0.1.0.post0
PyWavelets                1.2.0
pywin32                   303
pywinpty                  1.1.6
PyYAML                    6.0
pyzmq                     22.3.0
quantecon                 0.5.2
querystring-parser        1.2.4
rapidfuzz                 2.0.3
rasterio                  1.2.10
regex                     2021.11.10
requests                  2.26.0
requests-futures          1.0.0
Rtree                     0.9.7
s3transfer                0.3.7
scikit-learn              1.0.2
scikit-plot               0.3.7
scipy                     1.8.0
seaborn                   0.11.2
semver                    2.13.0
Send2Trash                1.8.0
setuptools                56.0.0
setuptools-scm            6.3.2
Shapely                   1.8.1.post1
simplejson                3.17.6
six                       1.16.0
sklearn                   0.0
smart-open                5.2.1
smmap                     5.0.0
snuggs                    1.4.7
soupsieve                 2.3.1
spacy                     3.2.1
spacy-legacy              3.0.8
spacy-loggers             1.0.1
spaghetti                 1.6.4
spglm                     1.0.8
spreg                     1.2.4
SQLAlchemy                1.4.29
sqlparse                  0.4.2
srsly                     2.4.2
stack-data                0.1.4
statsmodels               0.13.2
sympy                     1.9
tabulate                  0.8.9
tangled-up-in-unicode     0.1.0
tenacity                  8.0.1
terminado                 0.12.1
testpath                  0.5.0
textblob                  0.17.1
Theano-PyMC               1.1.2
thefuzz                   0.19.0
thinc                     8.0.13
threadpoolctl             3.0.0`

**Comment by martinfleis** [comment]
I think that you are running your code in a different environment that your `pip list` shows.

**Comment by lx12633036** [comment]
In this environment pip list show ,I have pygeos 0.12.0

**Comment by martinfleis** [comment]
Yes, in the environment in which you call `pip list`, you do indeed have pygeos. But I believe that your code actually uses a different environment than this one. You can try `import pygeos` directly in that environment. Maybe you'll get some useful error message.

**Comment by StanNieuwmans** [comment]
@lx12633036 @martinfleis Did anyone find a fix? I have exactly the same problem ;(
If I try to `import pygeos` I get this error: `DLL load failed while importing lib`

**Comment by lx12633036** [comment]
> 

nope,i can't import pygeos.So,I can't use this function！

**Comment by msumon94** [comment]
Hi,  
I am also facing same problem.    **any solution?**     geopandas.sjoin_nearest ( ) 

**error:** 
NotImplementedError: Currently, only PyGEOS >= 0.10.0 supports `nearest_all`. To use PyGEOS within GeoPandas, you need to install PyGEOS: 'conda install pygeos' or 'pip install pygeos'  

**here is the version **details****


SYSTEM INFO
-----------
python     : 3.7.13 (default, Apr 24 2022, 01:04:09)  [GCC 7.5.0]
executable : /usr/bin/python3
machine    : Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.6.2
GEOS lib   : /usr/lib/x86_64-linux-gnu/libgeos_c.so
GDAL       : 3.4.1
GDAL data dir: /usr/local/lib/python3.7/dist-packages/fiona/gdal_data
PROJ       : 8.1.1
PROJ data dir: /usr/local/lib/python3.7/dist-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.2
pandas     : 1.3.5
fiona      : 1.8.21
numpy      : 1.21.6
shapely    : 1.8.2
rtree      : None
pyproj     : 3.2.1
matplotlib : 3.2.2
mapclassify: None
geopy      : 1.17.0
psycopg2   : 2.7.6.1 (dt dec pq3 ext lo64)
geoalchemy2: None
pyarrow    : 6.0.1

**Comment by brendan-ward** [comment]
@msumon94 it looks like you do not have pygeos installed; as in the issues above, you need to have pygeos installed in 
the environment you are using.  It looks like you are not running inside a conda environment, but are on Linux, so `pip install pygeos` should work for you.

**Comment by brendan-ward** [comment]
@StanNieuwmans (sorry for the delayed response!) your issue looks different; you have installed `pygeos' but appear to be running into DLL loading issues.  You didn't list enough information about your environment for us to troubleshoot: what OS (looks like Windows?), how did you install things (pip vs conda), etc.

Perhaps reinstalling geopandas + pygeos inside a clean conda environment might work for you.

**Comment by msumon94** [comment]
@brendan-ward   Hi,
I appreciate your reply.  I used !pip install pygeos.  But still it is not working. 

I am running my code on google colab (jupyter notebook). 


import pandas as pd
!pip install geopandas
import geopandas as gpd
from shapely.geometry import Point, Polygon
%matplotlib inline   


!pip install pygeos 

lighting_with_crashes=gpd.sjoin_nearest(lighting_interchange, crash_interchange, how='left', max_distance=161)
lighting_with_crashes.head() 

then it shows: following error: 


NotImplementedError: Currently, only PyGEOS >= 0.10.0 supports `nearest_all`. To use PyGEOS within GeoPandas, you need to install PyGEOS: 'conda install pygeos' or 'pip install pygeos'




**My version** 

SYSTEM INFO
-----------
python     : 3.7.13 (default, Apr 24 2022, 01:04:09)  [GCC 7.5.0]
executable : /usr/bin/python3
machine    : Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.6.2
GEOS lib   : /usr/lib/x86_64-linux-gnu/libgeos_c.so
GDAL       : 3.4.1
GDAL data dir: /usr/local/lib/python3.7/dist-packages/fiona/gdal_data
PROJ       : 8.1.1
PROJ data dir: /usr/local/lib/python3.7/dist-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.2
pandas     : 1.3.5
fiona      : 1.8.21
numpy      : 1.21.6
shapely    : 1.8.2
rtree      : None
pyproj     : 3.2.1
matplotlib : 3.2.2
mapclassify: None
geopy      : 1.17.0
psycopg2   : 2.7.6.1 (dt dec pq3 ext lo64)
geoalchemy2: None
pyarrow    : 6.0.1
pygeos     : 0.12.0


**I also tried:**
 !pip install --upgrade pygeos==0.10.0
import pygeos

**Comment by brendan-ward** [comment]
@msumon94 The problem is the installation and execution order on Google collab; installing pygeos after importing geopandas means that geopandas does not detect pygeos when it starts up - even if you install / import pygeos after.  You need to restart your collab runtime first, then import pygeos, then geopandas, then import geopandas.

Here is a demo [notebook](https://gist.github.com/brendan-ward/245727ecaf511f742c0e3be9de8fa030) that shows pygeos now correctly available. (and a [broken one](https://gist.github.com/brendan-ward/6b72c68ef9a1af84565a89e3d8a249f0) demonstrating the error described above)

**Comment by msumon94** [comment]
@brendan-ward  thank you very much. Eventually, it worked. I really appreciate your support. But can you explain why installation order matter? 

Another thing: what is the unit of  max_distance for sjoin_nearest? Meter ? 

**my code**
lighting_with_crashes=geopandas.sjoin_nearest(lighting_interchange, crash_interchange, how='left', max_distance=161)

**here epsg for both datasets is 3857. Thats is why max_distance=161 means 161 meters. right?**

**Comment by brendan-ward** [comment]
To be clear, it is the installation _and_ import order.  When geopandas is imported, it checks to see if pygeos is available, and sets an internal flag that is used for all the operations that depend on pygeos.  If geopandas is imported before pygeos is installed, this flag is set to false; the interpreter then needs to be restarted to re-import geopandas now that pygeos is installed.

`max_distance` is in units of the crs of your GeoDataFrame.  Yes for EPSG:3857 the units are in meters.  Beware that 3857 (Web Mercator) is a projection for display of web maps and not appropriate for analysis (it distorts areas and lengths) - though choosing the correct projection is a more involved discussion outside the scope of this issue.

**Comment by msumon94** [comment]
@brendan-ward  thank you for your reply. 

Can you please tell me which projection (epsg) should be selected for this sjoin_nearest( ) function. I used epsg 3857, but it's not giving the correct result. Please note that  max_distance =161 which is in meter.

**Comment by brendan-ward** [comment]
@msumon94 The correct projection depends on the extent of your data and your goals / tradeoffs with respect to your analysis, and is off-topic for this issue.  I suggest an internet search to help find suggestions for your case, or to ask on the [Geopandas Gitter](https://gitter.im/geopandas/geopandas).  [This article](https://source.opennews.org/articles/choosing-right-map-projection/) might help.

> it's not giving the correct result

Can you please elaborate on what it is returning and how that is incorrect, so that we can identify if there is a bug?

**Comment by msumon94** [comment]
@brendan-ward  thank you very much

**Comment by ozzalel** [comment]
> > 
> 
> nope,i can't import pygeos.So,I can't use this function！

Hi, just ran into thie issue myself, including seeing pygeos on pip list and not seeing it on `geopandas.show_versions()`
I've installed pygeos using pip install, but I guess I've updated gepadnas using 
`conda install -c conda-forge geopandas=0.10`
The solution was pip uninstall pygeos and then installing pygeos using conda forge
`conda install pygeos --channel conda-forge` from the geopandas docs
after this process geopandas.show_versions() worked and sjoin worked

**Comment by sjabarnia** [comment]
I'm having the same problem, pygeos is installed but geopandas fails to use it.


> **SYSTEM INFO**
> -----------
> python     : 3.8.10 | packaged by conda-forge | (default, May 11 2021, 06:27:18)  [Clang 11.1.0 ]
> executable : ~/bin/python
> machine    : macOS-12.4-arm64-i386-64bit
> 
> GEOS, GDAL, PROJ INFO
> ---------------------
> GEOS       : 3.10.3
> GEOS lib   : ~/lib/libgeos_c.dylib
> GDAL       : 3.3.1
> GDAL data dir: ~/share/gdal
> PROJ       : 8.0.1
> PROJ data dir: ~/share/proj
> 
> PYTHON DEPENDENCIES
> -------------------
> geopandas  : 0.11.1
> numpy      : 1.21.6
> pandas     : 1.4.2
> pyproj     : 3.3.1
> shapely    : 1.8.0
> fiona      : 1.8.20
> geoalchemy2: 0.9.4
> geopy      : 2.1.0
> matplotlib : 3.5.1
> mapclassify: 2.4.3
> pygeos     : 0.10.2
> pyogrio    : None
> psycopg2   : 2.9.3 (dt dec pq3 ext lo64)
> pyarrow    : 5.0.0
> rtree      : 1.0.0

Following the steps in this [notebook](https://gist.github.com/brendan-ward/245727ecaf511f742c0e3be9de8fa030) :

```
import pandas as pd
import geopandas as gpd
import pygeos
print(f"pygeos version: {pygeos.__version__}")
print(f"geos version: {pygeos.geos_version}")
print(f"use pygeos? {gpd.options.use_pygeos}")
```

> pygeos version: 0.10.2
> geos version: (3, 9, 1)
> use pygeos? True

Still when trying to use sjoin_nearest, I get this error:

> NotImplementedError: Currently, only PyGEOS >= 0.10.0 supports `nearest_all`. To use PyGEOS within GeoPandas, you need to install PyGEOS: 'conda install pygeos' or 'pip install pygeos'

**Comment by martinfleis** [comment]
@sjabarnia have you tried the solutions described above?

**Comment by Coulorful** [comment]
> * [ ]  I have searched the [geopandas] tag on [StackOverflow](https://stackoverflow.com/questions/tagged/geopandas) and [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/geopandas) for similar questions.
> * [ ]  I have asked my usage related question on [StackOverflow](https://stackoverflow.com) or [GIS StackExhange](https://gis.stackexchange.com).
> 
> #### Question about geopandas
> **Note**: If you'd still like to submit a question, please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your question.
> 
> ```python
> # Your code here, if applicable
> [NotImplementedError: Currently, only PyGEOS >= 0.10.0 supports `nearest_all`. To use PyGEOS within GeoPandas, you need to install PyGEOS: 'conda install pygeos' or 'pip install pygeos']
> ```
> 
> But，my geopandas version is 0.10.2. And my pygeos version is 0.12.0

请问这个问题解决了吗？have sovled?

**Comment by martinfleis** [comment]
@Coulorful I am sorry, you will have to post your question in English.

**Comment by Coulorful** [comment]
> @Coulorful I am sorry, you will have to post your question in English.

Thanks for your reply. I have solved my problem, maybe due to pip install pyges ,  import pygeos from .lib import GEOSException # NOQA....ImportError: DLL load failed: The specified procedure could not be found.  Then I unintall it and install from [pygeos‑0.12.0‑cp39‑cp39‑win_amd64.whl]  finally it worked to import pygeos and  now I can use gpd.sjoin_nearest.
